contracts/src/everlong/EverlongPositions.sol

```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.20;
+
+import { IEverlongPositions } from "../interfaces/IEverlongPositions.sol";
+import { DoubleEndedQueue } from "openzeppelin/utils/structs/DoubleEndedQueue.sol";
+
+contract EverlongPositions is IEverlongPositions {
+    using DoubleEndedQueue for DoubleEndedQueue.Bytes32Deque;
+    DoubleEndedQueue.Bytes32Deque internal _positions;
```

Comment:
I'm not convinced we need a deque. I think we should start with a mapping and switch to this if necessary. I only see us needing to traverse backwards, if at all. We can do this with a queue.

Reply:
One case that seems like it would greatly benefit from a double-sided queue is when multiple longs are opened at different times in the same checkpoint interval. 

Without a pointer to the last item, the nieve approach for handling this would be to append a new position to the queue (im sure there's a better way but don't know it yet)... however this comes at the cost of more iterations when rebalancing is performed.

As a bonus,  a double-ended queue would be very handy to have in place should we choose to start targeting an avg. time to maturity.

Could totally be missing something, but it seems like the storage variables required for a single ended queue are identical to those used for a double ended queue.

[**Single-Ended Queue**](https://ethereum.stackexchange.com/questions/129668/how-to-efficiently-implement-a-fifo-array-queue-in-solidity)
```solidity
pragma solidity >=0.8.4;
contract Queue {
    mapping (uint256 => uint256 /* or any other type */ ) queue;
    uint256 first = 1;
    uint256 last = 1;
    function enqueue(uint256 data) public {
        last += 1;
        queue[uint256] = data;
    }

    function dequeue() public returns (uint256) {
        uint256 data;
        require(last > first);
        data = queue[first];
        delete queue[first];
        first += 1;
        return data
    }
    function length() public view returns (uint256) {
        return last - first;
    }
}
```

[**Double-Ended Queue**](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/structs/DoubleEndedQueue.sol#L29-L33)
```solidity
    struct Bytes32Deque {
        uint128 _begin;
        uint128 _end;
        mapping(uint128 index => bytes32) _data;
    }
    function pushBack(Bytes32Deque storage deque, bytes32 value) internal {
        unchecked {
            uint128 backIndex = deque._end;
            if (backIndex + 1 == deque._begin) Panic.panic(Panic.RESOURCE_ERROR);
            deque._data[backIndex] = value;
            deque._end = backIndex + 1;
        }
    }
    function popFront(Bytes32Deque storage deque) internal returns (bytes32 value) {
        unchecked {
            uint128 frontIndex = deque._begin;
            if (frontIndex == deque._end) Panic.panic(Panic.EMPTY_ARRAY_POP);
            value = deque._data[frontIndex];
            delete deque._data[frontIndex];
            deque._begin = frontIndex + 1;
        }
    }    
    function length(Bytes32Deque storage deque) internal view returns (uint256) {
        unchecked {
            return uint256(deque._end - deque._begin);
        }
    }
```
 
If the difference is only in implementing the methods for the double-ended queue, then why not use it?  Especially since OZ seems to have done a great job writing it for us 🤓 

FWIW I'm totally on board with the simplicity of the single sided approach and was initially going to go that route, just seemed  to me like the double-sided might come with some benefits at little-to-no cost.